<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <meta name="author" content="Chenghao Zhu">
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>File Format - moPepGen</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="../css/theme.css" />
  <link rel="stylesheet" href="../css/theme_extra.css" />
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
  <link href="../assets/_mkdocstrings.css" rel="stylesheet" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "File Format";
    var mkdocs_page_input_path = "file-format.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> moPepGen</a>
        
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../basics/">moPepGen Basics</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../quick-start/">Quick Start</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../vignette/">Vignette</a>
                    </li>
                </ul>
                <ul class="current">
                    <li class="toctree-l1 current"><a class="reference internal current" href="./">File Format</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#1-gene-variant-format">1 Gene Variant Format</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#11-file-metadata">1.1 File Metadata</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#12-point-mutation">1.2 Point Mutation</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#13-fusion">1.3 Fusion</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#14-alternative-splicing-site">1.4 Alternative Splicing Site</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#15-circrna">1.5 CircRNA</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#2-variant-peptide-fasta">2 Variant Peptide FASTA</a>
    </li>
    </ul>
                    </li>
                </ul>
                <p class="caption"><span class="caption-text">Commands</span></p>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../generate-index/">generateIndex</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../update-index/">updateIndex</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../parse-vep/">parseVEP</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../parse-reditools/">parseREDItools</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../parse-star-fusion/">parseSTARFusion</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../parse-fusion-catcher/">parseFusionCatcher</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../parse-arriba/">parseArriba</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../parse-rmats/">parseRMATS</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../parse-circexplorer/">parseCIRCexplorer</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../call-variant/">callVariant</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../call-novel-orf/">callNovelORF</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../call-alt-translation/">callAltTranslation</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../filter-fasta/">filterFasta</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../split-fasta/">splitFasta</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../merge-fasta/">mergeFasta</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../encode-fasta/">encodeFasta</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../decoy-fasta/">decoyFasta</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../summarize-fasta/">summarizeFasta</a>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">moPepGen</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>File Format</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/uclahs-cds/package-moPepGen/edit/master/docs/file-format.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  
  <hr/>
</div>

          <div role="main">
            <div class="section">
              
                <h1 id="file-structure-documentation">File Structure Documentation</h1>
<ul>
<li><a href="#file-structure-documentation">File Structure Documentation</a><ul>
<li><a href="#1-gene-variant-format">1 Gene Variant Format</a><ul>
<li><a href="#11-file-metadata">1.1 File Metadata</a></li>
<li><a href="#12-point-mutation">1.2 Point Mutation</a></li>
<li><a href="#13-fusion">1.3 Fusion</a></li>
<li><a href="#14-alternative-splicing-site">1.4 Alternative Splicing Site</a></li>
<li><a href="#15-circrna">1.5 CircRNA</a></li>
</ul>
</li>
<li><a href="#2-variant-peptide-fasta">2 Variant Peptide FASTA</a></li>
</ul>
</li>
</ul>
<h2 id="1-gene-variant-format">1 Gene Variant Format</h2>
<p>In moPepGen we are interested in finding variant peptides caused by combinations of different types of variants, including single nucleotide substitution, INDEL, RNA editing site, gene fusion and alternative splicing. We are also interested in non-coding RNA and circRNA with unreported ORF or start codon gained from mutation.</p>
<p>The different mutation events are called by different algorithms with varying output formats. In moPepGen, data type and tool-specific parsers convert variant data from different sources to a standardized VCF-like format. They are used in the <code>moPepGen callVariant</code> command to create the transcript variant graph and call variant peptides.</p>
<p>In moPepGen, we define the GVF (Gene Variant Format) file format, extended and modified from the <a href="https://samtools.github.io/hts-specs/VCFv4.2.pdf">VCF</a> file format, to represent the variant records. In a GVF file, each entry represents a variant associated with a transcript. The <code>CHROM</code> column is used to hold the gene ID, and the <code>POS</code> column indicates the variant position in reference to the transcript gene.</p>
<h3 id="11-file-metadata">1.1 File Metadata</h3>
<p>Each variant GVF file contains a metadata section, where each line starts with a double hashtag. The first line of the metadata must be the <code>fileformat</code> field in consistency with VCF standards. The line should read:</p>
<pre><code>##fileformat=VCFv4.2
</code></pre>
<p>moPepGen-specific GVF metadata starts on the second line. Each line is a key-value pair separated by an equal sign (<code>=</code>). Keys follow the <code>snake_case</code>. See examples below.</p>
<pre><code>##moPepGen_version=0.0.1
##parser=parseXXX
##reference_index=/path/to/reference-index
##genome_fasta=/path/to/genome.fasta
##annotation_gtf=/path/to/annotation.gtf
##source=SNP
</code></pre>
<ul>
<li><code>moPepGen_version</code>: the version number of moPepGen used to generate the variant file.</li>
<li><code>parser</code>: the parser used to create the variant file.</li>
<li><code>reference_index</code>: the path to the reference index used.</li>
<li><code>genome_fasta</code>: The genome fasta file used.</li>
<li><code>annotation_gtf</code>: The annotation GTF file used.</li>
<li><code>source</code>: The source of variants (e.g., SNP, SNV, Fusion). This value is used in the splitDatabase subcommand.</li>
</ul>
<p>If reference_index is not empty, <code>genome_fasta</code> and <code>annotation_gtf</code> refer to the files used to generate the index.</p>
<p>After the moPepGen metadata section, there should be a section for field information, that defines the fields used in either <code>ALT</code> or <code>INFO</code> column. For example:</p>
<pre><code>##INFO=&lt;ID=ID,Number=number,Type=type,Description=&quot;description&quot;&gt;
</code></pre>
<h3 id="12-point-mutation">1.2 Point Mutation</h3>
<p>Below is an example of a GVF file for point mutation, including single nucleotide substitution and small INDEL.</p>
<pre><code>##fileformat=VCFv4.2
##mopepgen_version=0.0.1
##parser=parseVEP
##reference_index=
##genome_fasta=
##annotation_gtf=
##source=SNP
##CHROM=&lt;Description='Gene ID'&gt;
##INFO=&lt;ID=TRANSCRIPT_ID,Number=1,Type=String,Description=&quot;Transcript ID&quot;&gt;
##INFO=&lt;ID=GENE_SYMBOL,Number=1,Type=String,Description=&quot;Gene Symbol&quot;&gt;
##INFO=&lt;ID=GENOMIC_POSITION,Number=1,Type=String,Description=&quot;Genomic Position&quot;&gt;
#CHROM    POS  ID           REF  ALT  QUAL  FILTER  INFO
ENSG0001  110  SNV-110-C-A  C    A    .     .       TRANSCRIPT_ID=ENST00011,ENST00012;GENE_SYMBOL=TP53;GENOMIC_POSITION=&quot;chr1:1000-1001&quot;
ENSG0002  210  SNV-210-T-A  T    A    .     .       TRANSCRIPT_ID=ENST00021,ENST00022;GENE_SYMBOL=EGFR;GENOMIC_POSITION=&quot;chr1:1000-1001&quot;
</code></pre>
<p>The <code>REF</code> and <code>ALT</code> must be explicit. The <code>INFO</code> column should contain the gene ID that the transcript belongs to. The <code>ID</code> column follows the pattern of '\<varant_type>-\<position>-\<ref>-\<alt>'.</p>
<h3 id="13-fusion">1.3 Fusion</h3>
<p>Below is an example of a GVF file for gene fusions.</p>
<pre><code>##fileformat=VCFv4.2
##mopepgen_version=0.0.1
##parser=parseXXX
##reference_index=
##genome_fasta=
##annotation_gtf=
##source=Fusion
##CHROM=&lt;Description='Gene ID'&gt;
##ALT=&lt;ID=FUSION,Description=&quot;Fusion&quot;&gt;
##INFO=&lt;ID=TRANSCRIP_ID,Number=&quot;+&quot;,Type=String,Description=&quot;5' Junction (Donor) Transcript ID&quot;&gt;
##INFO=&lt;ID=GENE_SYMBOL,Number=1,Type=String,Description=&quot;Gene Symbol&quot;&gt;
##INFO=&lt;ID=ACCEPTER_GENE_ID,Number=1,Type=String,Description=&quot;3' Junction (Accepter) Transcript's Gene ID&quot;&gt;
##INFO=&lt;ID=ACCEPTER_TRANSCRIPT_ID,Number=1,Type=String,Description=&quot;3' Junction (Accepter) Transcript's Transcript ID&quot;&gt;
##INFO=&lt;ID=ACCEPTER_POS,Number=1,Type=Integer,Description=&quot;Position of the break point of the accepter transcript&quot;&gt;
##INFO=&lt;ID=GENOMIC_POSITION,Number=1,Type=String,Description=&quot;Genomic Position&quot;&gt;
##INFO=&lt;ID=DONOR_GENOMIC_POSITION,Number=1,Type=String,Description=&quot;Genomic Position&quot;&gt;
##INFO=&lt;ID=DONOR_GENE_SYMBOL,Number=1,Type=String,Description=&quot;5' Junction (Donor) Gene Symbol&quot;&gt;
#CHROM    POS  ID                                 REF  ALT       QUAL  FILTER  INFO
ENSG0001  500  FUSION-ENSG0001:500-ENSG0011:1000  A    &lt;FUSION&gt;  .     .       TRANSCRIPT_ID=ENST0000;GENE_SYMBOL=SYMB1;ACCEPTER_GENE_ID=ENSG0010;ACCEPTER_TRANSCRIPT_ID=ENST0011;ACCEPTER_POSITION=1000;GENOMIC_POSITION=chr1:1000-1000;ACCEPTER_GENOMIC_POSITION=chr2:2000-2000;DONOR_GENE_SYMBOL=SYMB3
ENSG0002  500  FUSION-ENSG0002:500-ENSG0011:1000  A    &lt;FUSION&gt;  .     .       TRANSCRIPT_ID=ENST0000;GENE_SYMBOL=SYMB1;ACCEPTER_GENE_ID=ENSG0010;ACCEPTER_TRANSCRIPT_ID=ENST0011;ACCEPTER_POSITION=1000;GENOMIC_POSITION=chr1:1000-1000;ACCEPTER_GENOMIC_POSITION=chr2:2000-2000;DONOR_GENE_SYMBOL=SYMB3
ENSG0021  500  FUSION-ENSG0021:500-ENSG0031:1000  C    &lt;FUSION&gt;  .     .       TRANSCRIPT_ID=ENST0021;GENE_SYMBOL=SYMB2;ACCEPTER_GENE_ID=ENSG0030;ACCEPTER_TRANSCRIPT_ID=ENST0031;ACCEPTER_POSITION=1000;GENOMIC_POSITION=chr3:1000-1000;ACCEPTER_GENOMIC_POSITION=chr4:2000-2000;DONOR_GENE_SYMBOL=SYMB4
</code></pre>
<p>The <code>Info</code> column must contain the following fields:
+ <code>TRANSCRIPT_ID</code>: the transcript ID of the donor (upstream) transcript.
+ <code>ACCEPTER_GENE_ID</code>: the accepter (downstream) transcript gene ID.
+ <code>ACCEPTER_TRANSCRIPT_ID</code>: the accepter (downstream) transcript transcript ID.
+ <code>ACCEPTER_POSITION</code>: the position of the breakpoint of the ACCEPTER (downstream) transcript.
+ <code>GENOMIC_POSITION</code>: the genomic position of the donor (upstream) transcript, in the format of <code>&lt;chrom name&gt;:&lt;breakpoint&gt;-&lt;breakpoint&gt;</code>.
+ <code>ACCEPTER_GENOMIC_POSITION</code>: the genomic position of the accepter (downstream) transcript, in the format of <code>&lt;chrom name&gt;:&lt;breakpoint&gt;-&lt;breakpoint&gt;</code>.</p>
<h3 id="14-alternative-splicing-site">1.4 Alternative Splicing Site</h3>
<p>Alternative splicing sites called by <a href="http://rnaseq-mats.sourceforge.net/">rMATS</a> have five types, <em>i.e.</em> skipped exon (SE), alternative 5' splice site (A5SS), alternative 3' splice site (A3SS), mutually exclusive exons (MXE), and retained intron (RI). Each alternative splicing event can be represented as a deletion, insertion or substitution.</p>
<p>SE is called when an exon is skipped given its upstream and downstream exons. It is represented as an <strong>insertion</strong> when the target transcript from the GTF file contains the exon, or is represented as a <strong>deletion</strong> when the target transcript is annotated without the exon.</p>
<p>A5SS and A3SS are called when an exon has two splicing sites that can generate either a long or a short version of the exon. When the longer version is annotated in the given transcript, the variant is represented as a <strong>deletion</strong>, and an <strong>insertion</strong> is used when the shorter version is annotated.</p>
<p>MXE is represented as the substitution of one exon with another exon.</p>
<p>RI is represented as an insertion of the intron sequence.</p>
<pre><code>##fileformat=VCFv4.2
##mopepgen_version=0.0.1
##parser=parseRMATS
##reference_index=/path/to/reference/index
##genome_fasta=/path/to/genome.fasta
##annotation_gtf=/path/to/annotaion.gtf
##source=AlternativeSplicing
##CHROM=&lt;Description='Gene ID'&gt;
##INFO=&lt;ID=TRANSCRIPT_ID,Number=1,Type=String,Description=&quot;Transcript ID&quot;&gt;
##INFO=&lt;ID=START,Number=1,Type=Integer,Description=&quot;Start Position&quot;&gt;
##INFO=&lt;ID=END,Number=1,Type=Integer,Description=&quot;End Position&quot;&gt;
##INFO=&lt;ID=DONOR_START,Number=1,Type=Integer,Description=&quot;Donor Start Position&quot;&gt;
##INFO=&lt;ID=DONOR_END,Number=1,Type=Integer,Description=&quot;Donor End Position&quot;&gt;
##INFO=&lt;ID=GENE_SYMBOL,Number=1,Type=String,Description=&quot;Gene Symbol&quot;&gt;
##INFO=&lt;ID=GENOMIC_POSITION,Number=1,Type=String,Description=&quot;Genomic Position&quot;&gt;
#CHROM    POS  ID            REF  ALT    QUAL  FILTER  INFO
ENSG0001  110  SE-300        C    &lt;INS&gt;  .     .       TRANSCRIPT_ID=ENST00011;DONOR_GENE_ID=ENSG0005;DONOR_START=300;DONOR_END=400;GENE_SYMBOL=TP53;GENOMIC_POSITION=chr1:1000-1001
ENSG0002  210  A5SS-210      T    &lt;DEL&gt;  .     .       TRANSCRIPT_ID=ENST00021;DONOR_GENE_ID=ENSG0006;DONOR_START=210;DONOR_END=400;GENE_SYMBOL=EGFR;GENOMIC_POSITION=chr1:1000-1001
ENSG0003  115  A3SS-320      T    &lt;INS&gt;  .     .       TRANSCRIPT_ID=ENST00031;DONOR_GENE_ID=ENSG0007;DONOR_START=320;DONOR_END=380;GENE_SYMBOL=EGFR;GENOMIC_POSITION=chr1:1000-1001
ENSG0004  277  MXE-477-1103  T    &lt;SUB&gt;  .     .       TRANSCRIPT_ID=ENST00041;DONOR_GENE_ID=ENSG0008;DONOR_START=477;DONOR_END=582;DONOR_START=1103;DONOR_END=1228;GENE_SYMBOL=EGFR;GENOMIC_POSITION=chr1:1000-1001
</code></pre>
<p><strong>Examples:</strong></p>
<pre><code>ENSG0001    110 SE-300  C   &lt;INS&gt;   .   .   TRANSCRIPT_ID=ENST0001;DONOR_START=300;DONOR_END=400;GENE_SYMBOL=TP53;GENOMIC_POSITION=chr1:1000-1001
</code></pre>
<p>The line above represents an SE (skipped exon), that the sequence of 300-400 of the gene ENSG0001 is inserted into the transcript of ENST0001 at position 110. In this case, all transcripts of the gene in the annotation GTF don't contain this exon.</p>
<pre><code>ENSG0002    210 A5SS-210    T   &lt;DEL&gt;   .   .   TRANSCRIPT_ID=ENST0002;START=210;END=400;GENE_SYMBOL=EGFR;GENOMIC_POSITION=chr1:1000-1001
</code></pre>
<p>The line above represents an A5SS (alternative 5' splicing site), where the nucleotides from positions 210 to 400 of the transcript ENST0002 are deleted. This A5SS is represented as a <strong>deletion</strong> because all transcripts of the gene in the annotation GTF have the longer version of the exon.</p>
<pre><code>ENSG0003    115 MXE-320 T   &lt;INS&gt;   .   .   TRANSCRIPT_ID=ENST0003;START=320;END=380;GENE_SYMBOL=EGFR;GENOMIC_POSITION=chr1:1000-1001
</code></pre>
<p>The line above represents an MXE (mutually exclusive exon), where the exon at position 320-380 of the gene ENSG0003 is retained in the transcript ENST0003 and resulted as an insertion at position 115 of the transcript. This MXE is represented as an <strong>insertion</strong> because none of the transcripts of this gene has the first exon retained and the second spliced in the GTF, but this transcript has both exons retained.</p>
<pre><code>ENSG0004    277 MXE-477-1103    T   &lt;SUB&gt;   .   .   TRANSCRIPT_ID=ENST0004;START=477;END=582;DONOR_START=1103;DONOR_END=1228;GENE_SYMBOL=EGFR;GENOMIC_POSITION=chr1:1000-1001
</code></pre>
<p>This line above represents an MXE where the exon at position 447-582 in the gene (transcript ENST0004 position 277) is replaced with the exon at position 1103-1228 of the gene.</p>
<h3 id="15-circrna">1.5 CircRNA</h3>
<p>Circular RNAs are derived from back-spliced exons and introns. They exist as independent RNA molecules and have the potential to be translated into proteins. We are interested in finding the possible peptide sequences that could be the result of circRNA translation, with and without additional variants (SNP, INDEL, etc). In this case, circRNAs per se are rather new transcripts backbones than variants and are also recorded in GVF files in moPepGen. In such a GVF file, each row represents a circRNA, with the gene ID it is associated with, the start position at the gene coordinate, the offset and length of each segment, and their exon or intron indices. Normally each segment is an exon, but with intron-retained alternative splicing, they could be introns.</p>
<pre><code>##fileformat=VCFv4.2
##mopepgen_version=0.0.1
##parser=parseCIRCexplorer
##reference_index=/path/to/reference/index
##genome_fasta=/path/to/genome.fasta
##annotation_gtf=/path/to/annotaion.gtf
##CHROM=&lt;Description='Gene ID'&gt;
##INFO=&lt;ID=OFFSET,Number=+,Type=Integer,Description=&quot;Offsets of fragments (exons or introns)&quot;&gt;
##INFO=&lt;ID=LENGTH,Number=+,Type=Integer,Description=&quot;Length of fragments (exons or introns)&quot;&gt;
##INFO=&lt;ID=INTRON,Number=+,Type=Integer,Description=&quot;Indices of fragments that are introns&quot;&gt;
##INFO=&lt;ID=TRANSCRIPT,Number=1,Type=String,Description=&quot;Transcripts associated with this circRNA&quot;&gt;
##INFO=&lt;ID=GENE_SYMBOL,Number=1,Type=String,Description=&quot;Gene Symbol&quot;&gt;
##POS=&lt;Description=&quot;Gene coordinate of circRNA start&quot;&gt;
#CHROM    POS  ID                            REF  ALT  QUAL  FILTER  INFO
ENSG00000128408.9   0   CIRC-ENST00000614167.2-0:464    .   .   .   .   OFFSET=0,323;LENGTH=323,82;INTRON=;TRANSCRIPT_ID=ENST00000614167.2;GENE_SYMBOL=RIBC2;GENOMIC_POSITION=chr22:0:464
ENSG00000099949.21  0   CIRC-ENST00000642151.1-0:197    .   .   .   .   OFFSET=0,98;LENGTH=78,42;INTRON=;TRANSCRIPT_ID=ENST00000642151.1;GENE_SYMBOL=LZTR1;GENOMIC_POSITION=chr22:4980:5177
ENSG00000099949.21  78  CIRC-ENST00000642151.1-78:98    .   .   .   .   OFFSET=0;LENGTH=20;INTRON=0;TRANSCRIPT_ID=ENST00000642151.1;GENE_SYMBOL=LZTR1;GENOMIC_POSITION=chr22:5058:5078
</code></pre>
<p>circRNAs are not variants that are added to the transcript variant graph, thus the <code>REF</code> and <code>ALT</code> columns should be kept empty as ".". The <code>INFO</code> column must contain the following fields.</p>
<ul>
<li><strong><code>OFFSET</code></strong>: The offset of each fragment after the <code>start</code> position of the gene. Each segment can be either an exon or an intron.</li>
<li><strong><code>LENGTH</code></strong>: The length of each fragment.</li>
<li><strong><code>INTRON</code></strong>: The indices of fragments that are introns.</li>
<li><strong><code>TRANSCRIPT</code></strong> The transcript ID of a transcript that is able to generate this circRNA (e.g. contains all exons and introns of the circRNA).</li>
<li><strong><code>GENE_SYMBOL</code></strong> The name of the gene.</li>
</ul>
<p>circRNAs IDs follow the format CIRC-\<transcript_id>-<upstrea>:<downstream>, where \<transcript_id> is taken from the INFO column, and \<upstream> and \<downstream> represent the upstream and downstream gene coordinates of the backsplicing site. For example, <code>CIRC-ENST0001.1-78:135</code> refers to a circRNA derived from transcript ENST0001.1 with backsplicing occurring between gene positions 78 and 135.</p>
<h2 id="2-variant-peptide-fasta">2 Variant Peptide FASTA</h2>
<p>In moPepGen, the headers of the final output variant peptide FASTA contain the transcript IDs and variants associated with this variant peptide. The header of a peptide record starts with the transcript backbone ID and is followed by the variant IDs that it is associated with, separated by '|'. The variant IDs are defined in the GVF files. An ORF ID is added if the peptide is found from a novel ORF of the transcript. In some cases, several non-canonical peptides from the same transcript may share the same variants. This is most common in cases of peptide miscleavages. In addition, a frameshifting variant may cause multiple non-canonical peptides. An integer index is thus always added to the end of each header entry to resolve redundancies.</p>
<p>If the same peptide is found in multiple transcripts, all are documented as separate entries in the fasta header, separated by space.</p>
<pre><code>&gt;ENST0001|SNV-110-C-A|1
XXXXXXXXXXXXXXXXXX
&gt;ENST0002|SNV-210-T-A|SNV_220-G-C|1
XXXXXXXXXXXXXXXXXXXXXXXXXXXXX
&gt;ENST0003|INDEL-350-T-TACT|1 ENST0004|ENSG0004|SYMB4|SNV-55-C-G|1
XXXXXXXXXXXXXXXX
&gt;ENST0005|INDEL-110-CAA-A|1
XXXXXXXXXXXXXXXXXX
&gt;ENST0005|INDEL-110-CAA-A|2
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
</code></pre>
<p>For circRNA, the FASTA headers follow this style: <code>&lt;circRNA-ID&gt;|&lt;variant_id_1&gt;|...|&lt;variant_id_k&gt;|&lt;index&gt;</code></p>
<pre><code>&gt;CIRC-ENSG0001-15:68|1
XXXXXXXXXXXXX
&gt;CIRC-ENSG0001-153:285|SNV-110-C-A|2
XXXXXXXXXXXXX
</code></pre>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../generate-index/" class="btn btn-neutral float-right" title="generateIndex">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../vignette/" class="btn btn-neutral" title="Vignette"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/uclahs-cds/package-moPepGen/" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../vignette/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../generate-index/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme_extra.js" defer></script>
    <script src="../js/theme.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
