<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <meta name="author" content="Chenghao Zhu">
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>callVariant - moPepGen</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="../css/theme.css" />
  <link rel="stylesheet" href="../css/theme_extra.css" />
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
  <link href="../assets/_mkdocstrings.css" rel="stylesheet" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "callVariant";
    var mkdocs_page_input_path = "call-variant.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> moPepGen</a>
        
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../basics/">moPepGen Basics</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../quick-start/">Quick Start</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../vignette/">Vignette</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../file-format/">File Format</a>
                    </li>
                </ul>
                <p class="caption"><span class="caption-text">Commands</span></p>
                <ul class="current">
                    <li class="toctree-l1"><a class="reference internal" href="../generate-index/">generateIndex</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../update-index/">updateIndex</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../parse-vep/">parseVEP</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../parse-reditools/">parseREDItools</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../parse-star-fusion/">parseSTARFusion</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../parse-fusion-catcher/">parseFusionCatcher</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../parse-arriba/">parseArriba</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../parse-rmats/">parseRMATS</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../parse-circexplorer/">parseCIRCexplorer</a>
                    </li>
                    <li class="toctree-l1 current"><a class="reference internal current" href="./">callVariant</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#moPepGen.cli.call_variant_peptide">moPepGen.cli.call_variant_peptide</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#reference">Reference</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#codon-table-start-codon">Codon Table &amp; Start Codon</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#codon-table">Codon Table</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#start-codons">Start Codons</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#default">Default</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#usage">Usage</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#arguments">Arguments</a>
    </li>
    </ul>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../call-novel-orf/">callNovelORF</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../call-alt-translation/">callAltTranslation</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../filter-fasta/">filterFasta</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../split-fasta/">splitFasta</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../merge-fasta/">mergeFasta</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../encode-fasta/">encodeFasta</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../decoy-fasta/">decoyFasta</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../summarize-fasta/">summarizeFasta</a>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">moPepGen</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
        
          <li>Commands &raquo;</li>
        
      
    
    <li>callVariant</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/uclahs-cds/package-moPepGen/edit/master/docs/call-variant.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  
  <hr/>
</div>

          <div role="main">
            <div class="section">
              
                <h1 id="callvariant">callVariant</h1>


  <div class="doc doc-object doc-module">

<a id="moPepGen.cli.call_variant_peptide"></a>
    <div class="doc doc-contents first">

      <p><code>callVariant</code> is the core of moPepGen. It takes multiple GVF files, generated
by any moPepGen parser, and calls variant peptides caused by genomic variants
using a graph-based algorithm. For any transcript, it creates a three-frame
transcript variant graph by incorporating all variants from any sources (SNV,
INDEL, fusion, alternative splicing, RNA editing, and circRNA). The transcript
variant graph is then translated into a peptide variant graph, followed by
converting to a cleavage graph based on the enzymatic cleavage rule. The
variant peptide graph is than used to call for variant peptides that contains
at least one variant, and do not present in the canonical peptide pool.</p>


    </div>

  </div>

<h2 id="reference">Reference</h2>
<p>Reference data, incluiding reference genome, genome annotation, and protein coding translation are required. There are two ways of specifying reference data:</p>
<ol>
<li>Using the index dir created by the <a href="/generate-index"><code>generateIndex</code></a> command.</li>
<li>Specifying each reference files needed.</li>
</ol>
<p>1 is highly recommended as it is faster and helps you ensure that the same reference data are used across the project.</p>
<h2 id="codon-table-start-codon">Codon Table &amp; Start Codon</h2>
<h3 id="codon-table">Codon Table</h3>
<p>The NCBI standard codon table is used by default, which is used for the majority of nulcear gene translation in eukaryote cells. See <a href="https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi">here</a> for a complete list of NCBI codon tables.</p>
<p>The default codon table can be override using <code>--codon-table</code>. For example:</p>
<pre><code class="language-shell">--codon-table â€™Ciliate Nuclear'
</code></pre>
<p>The <code>--chr-codon-table</code> can be used to specify the codon table used for a specific chomosome. The example below uses the 'Vertebrate Mitochondrial' (SGC1) codon table for genes from the mitochondria chomosome, and the standard codon table otherwise.</p>
<pre><code class="language-shell">--codon-table Standard \
--chr-codon-table 'chrM:SGC1'
</code></pre>
<h3 id="start-codons">Start Codons</h3>
<p>Stard codons usually do not need to be specified. The standard start codon <code>ATG</code> is used by default, and it is translated as Methionine as start codon and in elongation. However, in some cases, for example, mitochondria, <code>ATA</code> and <code>ATT</code> may also be used as start codon. While <code>ATT</code> is translatted into Isoleucine during elongation, Methionine is still used as start codon.</p>
<p>Similar to codon table, the default codon table can be override using <code>--start-codons</code>.</p>
<pre><code class="language-shell">--start-codons ATG
</code></pre>
<p>The <code>--chr-start-codon</code> can also be used to assign start codons to a specific chomosome. The example below assigns <code>ATG</code>, <code>ATA</code>, and <code>ATT</code> to the mitochondrial chromosome.</p>
<pre><code class="language-shell">--chr-start-codons 'chrM:ATG,ATA,ATT'
</code></pre>
<h3 id="default">Default</h3>
<p>The chromosome names must be specified correctly, same as what used in the genome fasta and annotation GTF file. By default, moPepGen infers the reference source of the annotation (<em>i.e.</em>, GENCODE or EMSEMBL), and uses the 'SGC1' codon table for mitochondirla chromosome. So the default is equivalent to:</p>
<pre><code class="language-shell">--reference-source GENCODE \
--codon-table Standard \
--chr-codon-table 'chrM:SGC1' \
--start-codons 'ATG' \
--chr-start-codons 'chrM:ATG,ATA,ATT
</code></pre>
<p>or</p>
<pre><code class="language-shell">--reference-source ENSEMBL \
--codon-table Standard \
--chr-codon-table 'MT:SGC1' \
--start-codons 'ATG' \
--chr-start-codons 'MT:ATG,ATA,ATT
</code></pre>
<h2 id="usage">Usage</h2>
<pre><code>usage: moPepGen callVariant [-h] -i ['&lt;files&gt;'] [['&lt;files&gt;'] ...] -o &lt;file&gt;
                            [--graph-output-dir &lt;file&gt;]
                            [--max-adjacent-as-mnv MAX_ADJACENT_AS_MNV]
                            [--selenocysteine-termination]
                            [--w2f-reassignment] [--backsplicing-only]
                            [--coding-novel-orf]
                            [--max-variants-per-node &lt;number&gt; [&lt;number&gt; ...]]
                            [--additional-variants-per-misc &lt;number&gt; [&lt;number&gt; ...]]
                            [--in-bubble-cap-step-down &lt;number&gt;]
                            [--min-nodes-to-collapse &lt;number&gt;]
                            [--naa-to-collapse &lt;number&gt;]
                            [--noncanonical-transcripts]
                            [--timeout-seconds TIMEOUT_SECONDS]
                            [--threads &lt;number&gt;] [--skip-failed] [-g &lt;file&gt;]
                            [-a &lt;file&gt;] [--reference-source {GENCODE,ENSEMBL}]
                            [--codon-table {Standard,Alternative Flatworm Mitochondrial,Trematode Mitochondrial,SGC8,SGC3,Protozoan Mitochondrial,Ciliate Nuclear,Gracilibacteria,Spiroplasma,Dasycladacean Nuclear,Invertebrate Mitochondrial,Balanophoraceae Plastid,Peritrich Nuclear,Mesodinium Nuclear,SGC5,Candidate Division SR1,Blastocrithidia Nuclear,SGC1,Bacterial,Alternative Yeast Nuclear,Yeast Mitochondrial,Scenedesmus obliquus Mitochondrial,Plant Plastid,Flatworm Mitochondrial,SGC2,Archaeal,Mycoplasma,Euplotid Nuclear,SGC9,Mold Mitochondrial,Thraustochytrium Mitochondrial,Hexamita Nuclear,Coelenterate Mitochondrial,Chlorophycean Mitochondrial,Pachysolen tannophilus Nuclear,Ascidian Mitochondrial,SGC0,Blepharisma Macronuclear,Karyorelict Nuclear,SGC4,Echinoderm Mitochondrial,Condylostoma Nuclear,Vertebrate Mitochondrial,Pterobranchia Mitochondrial,Cephalodiscidae Mitochondrial}]
                            [--chr-codon-table [CHR_CODON_TABLE ...]]
                            [--start-codons [START_CODONS ...]]
                            [--chr-start-codons [CHR_START_CODONS ...]]
                            [-p &lt;file&gt;] [--invalid-protein-as-noncoding]
                            [--index-dir [&lt;file&gt;]] [-c &lt;value&gt;]
                            [--cleavage-exception &lt;value&gt;] [-m &lt;number&gt;]
                            [-w &lt;number&gt;] [-l &lt;number&gt;] [-x &lt;number&gt;]
                            [--debug-level &lt;value|number&gt;] [-q]

Genomic variant data must be generated by one of the moPepGenparsers. See
moPepGen --help

options:
  -h, --help            show this help message and exit
  -i ['&lt;files&gt;'] [['&lt;files&gt;'] ...], --input-path ['&lt;files&gt;'] [['&lt;files&gt;'] ...]
                        File path to GVF files. Must be generated by any of
                        the moPepGen parsers. Can take multiple files. Valid
                        formats: ['.gvf'] (default: None)
  -o &lt;file&gt;, --output-path &lt;file&gt;
                        File path to the output file. Valid formats:
                        ['.fasta', '.fa'] (default: None)
  --graph-output-dir &lt;file&gt;
                        Directory path that graph data are saved to. Graph
                        data are not saved if this is not given. (default:
                        None)
  --max-adjacent-as-mnv MAX_ADJACENT_AS_MNV
                        Max number of adjacent variants that should be merged.
                        (default: 2)
  --selenocysteine-termination
                        Include peptides of selenoproteins that the UGA is
                        treated as termination instead of Sec. (default:
                        False)
  --w2f-reassignment    Include peptides with W &gt; F (Tryptophan to
                        Phenylalanine) reassignment. (default: False)
  --backsplicing-only   For circRNA, only keep noncanonical peptides spaning
                        the backsplicing site. (default: False)
  --coding-novel-orf    Find alternative start site for coding transcripts.
                        (default: False)
  --max-variants-per-node &lt;number&gt; [&lt;number&gt; ...]
                        Maximal number of variants per node. Useful when local
                        regions are heavily mutated. When creating the
                        cleavage graph, nodes containing variants larger than
                        this value are skipped. Set -1 to disable. (default:
                        [7])
  --additional-variants-per-misc &lt;number&gt; [&lt;number&gt; ...]
                        Additional variants allowed for every miscleavage.
                        Used with --max-variants-per-node to handle
                        hypermutated regions. Set -1 to disable. (default:
                        [2])
  --in-bubble-cap-step-down &lt;number&gt;
                        Default step down for in-bubble variant caps.
                        (default: 0)
  --min-nodes-to-collapse &lt;number&gt;
                        When making the cleavage graph, the minimal number of
                        nodes to trigger pop collapse. (default: 30)
  --naa-to-collapse &lt;number&gt;
                        The number of bases used for pop collapse. (default:
                        5)
  --noncanonical-transcripts
                        Process only noncanonical transcripts of fusion
                        transcripts and circRNA. Canonical transcripts are
                        skipped. (default: False)
  --timeout-seconds TIMEOUT_SECONDS
                        Timeout in seconds for each transcript. (default:
                        1800)
  --threads &lt;number&gt;    Set number of worker processes to use. (default: 1)
  --skip-failed         When set, the failed records will be skipped.
                        (default: False)
  --debug-level &lt;value|number&gt;
                        Debug level. (default: INFO)
  -q, --quiet           Quiet (default: False)

Reference Files:
  -g &lt;file&gt;, --genome-fasta &lt;file&gt;
                        Path to the genome assembly FASTA file. Only ENSEMBL
                        and GENCODE are supported. Its version must be the
                        same as the annotation GTF and proteome FASTA
                        (default: None)
  -a &lt;file&gt;, --annotation-gtf &lt;file&gt;
                        Path to the annotation GTF file. Only ENSEMBL and
                        GENCODE are supported. Its version must be the same as
                        the genome and proteome FASTA. (default: None)
  --reference-source {GENCODE,ENSEMBL}
                        Source of reference genome and annotation. (default:
                        None)
  --codon-table {Standard,Alternative Flatworm Mitochondrial,Trematode Mitochondrial,SGC8,SGC3,Protozoan Mitochondrial,Ciliate Nuclear,Gracilibacteria,Spiroplasma,Dasycladacean Nuclear,Invertebrate Mitochondrial,Balanophoraceae Plastid,Peritrich Nuclear,Mesodinium Nuclear,SGC5,Candidate Division SR1,Blastocrithidia Nuclear,SGC1,Bacterial,Alternative Yeast Nuclear,Yeast Mitochondrial,Scenedesmus obliquus Mitochondrial,Plant Plastid,Flatworm Mitochondrial,SGC2,Archaeal,Mycoplasma,Euplotid Nuclear,SGC9,Mold Mitochondrial,Thraustochytrium Mitochondrial,Hexamita Nuclear,Coelenterate Mitochondrial,Chlorophycean Mitochondrial,Pachysolen tannophilus Nuclear,Ascidian Mitochondrial,SGC0,Blepharisma Macronuclear,Karyorelict Nuclear,SGC4,Echinoderm Mitochondrial,Condylostoma Nuclear,Vertebrate Mitochondrial,Pterobranchia Mitochondrial,Cephalodiscidae Mitochondrial}
                        Codon table. Defaults to &quot;Standard&quot;. Supported codon
                        tables: {'Standard', 'Alternative Flatworm
                        Mitochondrial', 'Trematode Mitochondrial', 'SGC8',
                        'SGC3', 'Protozoan Mitochondrial', 'Ciliate Nuclear',
                        'Gracilibacteria', 'Spiroplasma', 'Dasycladacean
                        Nuclear', 'Invertebrate Mitochondrial',
                        'Balanophoraceae Plastid', 'Peritrich Nuclear',
                        'Mesodinium Nuclear', 'SGC5', 'Candidate Division
                        SR1', 'Blastocrithidia Nuclear', 'SGC1', 'Bacterial',
                        'Alternative Yeast Nuclear', 'Yeast Mitochondrial',
                        'Scenedesmus obliquus Mitochondrial', 'Plant Plastid',
                        'Flatworm Mitochondrial', 'SGC2', 'Archaeal',
                        'Mycoplasma', 'Euplotid Nuclear', 'SGC9', 'Mold
                        Mitochondrial', 'Thraustochytrium Mitochondrial',
                        'Hexamita Nuclear', 'Coelenterate Mitochondrial',
                        'Chlorophycean Mitochondrial', 'Pachysolen tannophilus
                        Nuclear', 'Ascidian Mitochondrial', 'SGC0',
                        'Blepharisma Macronuclear', 'Karyorelict Nuclear',
                        'SGC4', 'Echinoderm Mitochondrial', 'Condylostoma
                        Nuclear', 'Vertebrate Mitochondrial', 'Pterobranchia
                        Mitochondrial', 'Cephalodiscidae Mitochondrial'}
                        (default: Standard)
  --chr-codon-table [CHR_CODON_TABLE ...]
                        Chromosome specific codon table. Must be specified in
                        the format of &quot;chrM:SGC1&quot;, where &quot;chrM&quot; is the
                        chromosome name and &quot;SGC1&quot; is the codon table to use
                        to translate genes on chrM. Supported codon tables:
                        {'Standard', 'Alternative Flatworm Mitochondrial',
                        'Trematode Mitochondrial', 'SGC8', 'SGC3', 'Protozoan
                        Mitochondrial', 'Ciliate Nuclear', 'Gracilibacteria',
                        'Spiroplasma', 'Dasycladacean Nuclear', 'Invertebrate
                        Mitochondrial', 'Balanophoraceae Plastid', 'Peritrich
                        Nuclear', 'Mesodinium Nuclear', 'SGC5', 'Candidate
                        Division SR1', 'Blastocrithidia Nuclear', 'SGC1',
                        'Bacterial', 'Alternative Yeast Nuclear', 'Yeast
                        Mitochondrial', 'Scenedesmus obliquus Mitochondrial',
                        'Plant Plastid', 'Flatworm Mitochondrial', 'SGC2',
                        'Archaeal', 'Mycoplasma', 'Euplotid Nuclear', 'SGC9',
                        'Mold Mitochondrial', 'Thraustochytrium
                        Mitochondrial', 'Hexamita Nuclear', 'Coelenterate
                        Mitochondrial', 'Chlorophycean Mitochondrial',
                        'Pachysolen tannophilus Nuclear', 'Ascidian
                        Mitochondrial', 'SGC0', 'Blepharisma Macronuclear',
                        'Karyorelict Nuclear', 'SGC4', 'Echinoderm
                        Mitochondrial', 'Condylostoma Nuclear', 'Vertebrate
                        Mitochondrial', 'Pterobranchia Mitochondrial',
                        'Cephalodiscidae Mitochondrial'}. By default, &quot;SGC1&quot;
                        is assigned to mitochondrial chromosomes. (default:
                        [])
  --start-codons [START_CODONS ...]
                        Default start codon(s) to use for novel ORF
                        translation. Defaults to [&quot;ATG&quot;]. (default: ['ATG'])
  --chr-start-codons [CHR_START_CODONS ...]
                        Chromosome specific start codon(s). For example,
                        &quot;chrM:ATG,ATA,ATT&quot;.By defualt, mitochondrial
                        chromosome name is automatically inferred andstart
                        codon &quot;ATG&quot;, &quot;ATA&quot;, &quot;ATT&quot;, &quot;ATC&quot; and &quot;GTG&quot; are
                        assigned to it. (default: [])
  -p &lt;file&gt;, --proteome-fasta &lt;file&gt;
                        Path to the translated protein sequence FASTA file.
                        Only ENSEMBL and GENCODE are supported. Its version
                        must be the same as genome FASTA and annotation GTF.
                        (default: None)
  --invalid-protein-as-noncoding
                        Treat any transcript that the protein sequence is
                        invalid ( contains the * symbol) as noncoding.
                        (default: False)
  --index-dir [&lt;file&gt;]  Path to the directory of index files generated by
                        moPepGen generateIndex. If given, --genome-fasta,
                        --proteome-fasta and --anntotation-gtf will be
                        ignored. (default: None)

Cleavage Parameters:
  -c &lt;value&gt;, --cleavage-rule &lt;value&gt;
                        Enzymatic cleavage rule. (default: trypsin)
  --cleavage-exception &lt;value&gt;
                        Enzymatic cleavage exception. (default: auto)
  -m &lt;number&gt;, --miscleavage &lt;number&gt;
                        Number of cleavages to allow per non-canonical
                        peptide. (default: 2)
  -w &lt;number&gt;, --min-mw &lt;number&gt;
                        The minimal molecular weight of the non-canonical
                        peptides. (default: 500.0)
  -l &lt;number&gt;, --min-length &lt;number&gt;
                        The minimal length of non-canonical peptides,
                        inclusive. (default: 7)
  -x &lt;number&gt;, --max-length &lt;number&gt;
                        The maximum length of non-canonical peptides,
                        inclusive. (default: 25)

</code></pre>
<h2 id="arguments">Arguments</h2>
<p><strong>-h, --help</strong><br/></p>
<p style="margin-left: 15px;">
show this help message and exit


</p>

<p><strong>-i, --input-path</strong> <strong><em>[&#39;&lt;files&gt;&#39;]</em></strong> <em>Path</em><br/></p>
<p style="margin-left: 15px;">
File path to GVF files. Must be generated by any of the moPepGen parsers. Can take multiple files. Valid formats: ['.gvf']


</p>

<p><strong>-o, --output-path</strong> <strong><em>&lt;file&gt;</em></strong> <em>Path</em><br/></p>
<p style="margin-left: 15px;">
File path to the output file. Valid formats: ['.fasta', '.fa']


</p>

<p><strong>--graph-output-dir</strong> <strong><em>&lt;file&gt;</em></strong> <em>Path</em><br/></p>
<p style="margin-left: 15px;">
Directory path that graph data are saved to. Graph data are not saved if this is not given.


</p>

<p><strong>--max-adjacent-as-mnv</strong> <em>int</em><br/></p>
<p style="margin-left: 15px;">
Max number of adjacent variants that should be merged.
 int <br/>Default: 2 

</p>

<p><strong>--selenocysteine-termination</strong><br/></p>
<p style="margin-left: 15px;">
Include peptides of selenoproteins that the UGA is treated as termination instead of Sec.
  <br/>Default: False 

</p>

<p><strong>--w2f-reassignment</strong><br/></p>
<p style="margin-left: 15px;">
Include peptides with W > F (Tryptophan to Phenylalanine) reassignment.
  <br/>Default: False 

</p>

<p><strong>--backsplicing-only</strong><br/></p>
<p style="margin-left: 15px;">
For circRNA, only keep noncanonical peptides spaning the backsplicing site.
  <br/>Default: False 

</p>

<p><strong>--coding-novel-orf</strong><br/></p>
<p style="margin-left: 15px;">
Find alternative start site for coding transcripts.
  <br/>Default: False 

</p>

<p><strong>--max-variants-per-node</strong> <strong><em>&lt;number&gt;</em></strong> <em>int</em><br/></p>
<p style="margin-left: 15px;">
Maximal number of variants per node. Useful when local regions are heavily mutated. When creating the cleavage graph, nodes containing variants larger than this value are skipped. Set -1 to disable.
 int <br/>Default: [7] 

</p>

<p><strong>--additional-variants-per-misc</strong> <strong><em>&lt;number&gt;</em></strong> <em>int</em><br/></p>
<p style="margin-left: 15px;">
Additional variants allowed for every miscleavage. Used with --max-variants-per-node to handle hypermutated regions. Set -1 to disable.
 int <br/>Default: [2] 

</p>

<p><strong>--in-bubble-cap-step-down</strong> <strong><em>&lt;number&gt;</em></strong> <em>int</em><br/></p>
<p style="margin-left: 15px;">
Default step down for in-bubble variant caps.
 int <br/>Default: 0 

</p>

<p><strong>--min-nodes-to-collapse</strong> <strong><em>&lt;number&gt;</em></strong> <em>int</em><br/></p>
<p style="margin-left: 15px;">
When making the cleavage graph, the minimal number of nodes to trigger pop collapse.
 int <br/>Default: 30 

</p>

<p><strong>--naa-to-collapse</strong> <strong><em>&lt;number&gt;</em></strong> <em>int</em><br/></p>
<p style="margin-left: 15px;">
The number of bases used for pop collapse.
 int <br/>Default: 5 

</p>

<p><strong>--noncanonical-transcripts</strong><br/></p>
<p style="margin-left: 15px;">
Process only noncanonical transcripts of fusion transcripts and circRNA. Canonical transcripts are skipped.
  <br/>Default: False 

</p>

<p><strong>--timeout-seconds</strong> <em>int</em><br/></p>
<p style="margin-left: 15px;">
Timeout in seconds for each transcript.
 int <br/>Default: 1800 

</p>

<p><strong>--threads</strong> <strong><em>&lt;number&gt;</em></strong> <em>int</em><br/></p>
<p style="margin-left: 15px;">
Set number of worker processes to use.
 int <br/>Default: 1 

</p>

<p><strong>--skip-failed</strong><br/></p>
<p style="margin-left: 15px;">
When set, the failed records will be skipped.
  <br/>Default: False 

</p>

<p><strong>-g, --genome-fasta</strong> <strong><em>&lt;file&gt;</em></strong> <em>Path</em><br/></p>
<p style="margin-left: 15px;">
Path to the genome assembly FASTA file. Only ENSEMBL and GENCODE are supported. Its version must be the same as the annotation GTF and proteome FASTA


</p>

<p><strong>-a, --annotation-gtf</strong> <strong><em>&lt;file&gt;</em></strong> <em>Path</em><br/></p>
<p style="margin-left: 15px;">
Path to the annotation GTF file. Only ENSEMBL and GENCODE are supported. Its version must be the same as the genome and proteome FASTA.


</p>

<p><strong>--reference-source</strong> <em>str</em><br/></p>
<p style="margin-left: 15px;">
Source of reference genome and annotation.

 <br/>Choices: ['GENCODE', 'ENSEMBL'] 
</p>

<p><strong>--codon-table</strong> <em>str</em><br/></p>
<p style="margin-left: 15px;">
Codon table. Defaults to "Standard". Supported codon tables: {'Standard', 'Alternative Flatworm Mitochondrial', 'Trematode Mitochondrial', 'SGC8', 'SGC3', 'Protozoan Mitochondrial', 'Ciliate Nuclear', 'Gracilibacteria', 'Spiroplasma', 'Dasycladacean Nuclear', 'Invertebrate Mitochondrial', 'Balanophoraceae Plastid', 'Peritrich Nuclear', 'Mesodinium Nuclear', 'SGC5', 'Candidate Division SR1', 'Blastocrithidia Nuclear', 'SGC1', 'Bacterial', 'Alternative Yeast Nuclear', 'Yeast Mitochondrial', 'Scenedesmus obliquus Mitochondrial', 'Plant Plastid', 'Flatworm Mitochondrial', 'SGC2', 'Archaeal', 'Mycoplasma', 'Euplotid Nuclear', 'SGC9', 'Mold Mitochondrial', 'Thraustochytrium Mitochondrial', 'Hexamita Nuclear', 'Coelenterate Mitochondrial', 'Chlorophycean Mitochondrial', 'Pachysolen tannophilus Nuclear', 'Ascidian Mitochondrial', 'SGC0', 'Blepharisma Macronuclear', 'Karyorelict Nuclear', 'SGC4', 'Echinoderm Mitochondrial', 'Condylostoma Nuclear', 'Vertebrate Mitochondrial', 'Pterobranchia Mitochondrial', 'Cephalodiscidae Mitochondrial'}
 str <br/>Default: Standard 
 <br/>Choices: {'Standard', 'Alternative Flatworm Mitochondrial', 'Trematode Mitochondrial', 'SGC8', 'SGC3', 'Protozoan Mitochondrial', 'Ciliate Nuclear', 'Gracilibacteria', 'Spiroplasma', 'Dasycladacean Nuclear', 'Invertebrate Mitochondrial', 'Balanophoraceae Plastid', 'Peritrich Nuclear', 'Mesodinium Nuclear', 'SGC5', 'Candidate Division SR1', 'Blastocrithidia Nuclear', 'SGC1', 'Bacterial', 'Alternative Yeast Nuclear', 'Yeast Mitochondrial', 'Scenedesmus obliquus Mitochondrial', 'Plant Plastid', 'Flatworm Mitochondrial', 'SGC2', 'Archaeal', 'Mycoplasma', 'Euplotid Nuclear', 'SGC9', 'Mold Mitochondrial', 'Thraustochytrium Mitochondrial', 'Hexamita Nuclear', 'Coelenterate Mitochondrial', 'Chlorophycean Mitochondrial', 'Pachysolen tannophilus Nuclear', 'Ascidian Mitochondrial', 'SGC0', 'Blepharisma Macronuclear', 'Karyorelict Nuclear', 'SGC4', 'Echinoderm Mitochondrial', 'Condylostoma Nuclear', 'Vertebrate Mitochondrial', 'Pterobranchia Mitochondrial', 'Cephalodiscidae Mitochondrial'} 
</p>

<p><strong>--chr-codon-table</strong> <em>str</em><br/></p>
<p style="margin-left: 15px;">
Chromosome specific codon table. Must be specified in the format of "chrM:SGC1", where "chrM" is the chromosome name and "SGC1" is the codon table to use to translate genes on chrM. Supported codon tables: {'Standard', 'Alternative Flatworm Mitochondrial', 'Trematode Mitochondrial', 'SGC8', 'SGC3', 'Protozoan Mitochondrial', 'Ciliate Nuclear', 'Gracilibacteria', 'Spiroplasma', 'Dasycladacean Nuclear', 'Invertebrate Mitochondrial', 'Balanophoraceae Plastid', 'Peritrich Nuclear', 'Mesodinium Nuclear', 'SGC5', 'Candidate Division SR1', 'Blastocrithidia Nuclear', 'SGC1', 'Bacterial', 'Alternative Yeast Nuclear', 'Yeast Mitochondrial', 'Scenedesmus obliquus Mitochondrial', 'Plant Plastid', 'Flatworm Mitochondrial', 'SGC2', 'Archaeal', 'Mycoplasma', 'Euplotid Nuclear', 'SGC9', 'Mold Mitochondrial', 'Thraustochytrium Mitochondrial', 'Hexamita Nuclear', 'Coelenterate Mitochondrial', 'Chlorophycean Mitochondrial', 'Pachysolen tannophilus Nuclear', 'Ascidian Mitochondrial', 'SGC0', 'Blepharisma Macronuclear', 'Karyorelict Nuclear', 'SGC4', 'Echinoderm Mitochondrial', 'Condylostoma Nuclear', 'Vertebrate Mitochondrial', 'Pterobranchia Mitochondrial', 'Cephalodiscidae Mitochondrial'}. By default, "SGC1" is assigned to mitochondrial chromosomes.
 str <br/>Default: [] 

</p>

<p><strong>--start-codons</strong> <em>str</em><br/></p>
<p style="margin-left: 15px;">
Default start codon(s) to use for novel ORF translation. Defaults to ["ATG"].
 str <br/>Default: ['ATG'] 

</p>

<p><strong>--chr-start-codons</strong> <em>str</em><br/></p>
<p style="margin-left: 15px;">
Chromosome specific start codon(s). For example, "chrM:ATG,ATA,ATT".By defualt, mitochondrial chromosome name is automatically inferred andstart codon "ATG", "ATA", "ATT", "ATC" and "GTG" are assigned to it.
 str <br/>Default: [] 

</p>

<p><strong>-p, --proteome-fasta</strong> <strong><em>&lt;file&gt;</em></strong> <em>Path</em><br/></p>
<p style="margin-left: 15px;">
Path to the translated protein sequence FASTA file. Only ENSEMBL and GENCODE are supported. Its version must be the same as genome FASTA and annotation GTF.


</p>

<p><strong>--invalid-protein-as-noncoding</strong><br/></p>
<p style="margin-left: 15px;">
Treat any transcript that the protein sequence is invalid ( contains the * symbol) as noncoding.
  <br/>Default: False 

</p>

<p><strong>--index-dir</strong> <strong><em>&lt;file&gt;</em></strong> <em>Path</em><br/></p>
<p style="margin-left: 15px;">
Path to the directory of index files generated by moPepGen generateIndex. If given, --genome-fasta, --proteome-fasta and --anntotation-gtf will be ignored.


</p>

<p><strong>-c, --cleavage-rule</strong> <strong><em>&lt;value&gt;</em></strong> <em>str</em><br/></p>
<p style="margin-left: 15px;">
Enzymatic cleavage rule.
 str <br/>Default: trypsin 
 <br/>Choices: ['arg-c', 'asp-n', 'bnps-skatole', 'caspase 1', 'caspase 2', 'caspase 3', 'caspase 4', 'caspase 5', 'caspase 6', 'caspase 7', 'caspase 8', 'caspase 9', 'caspase 10', 'chymotrypsin high specificity', 'chymotrypsin low specificity', 'clostripain', 'cnbr', 'enterokinase', 'factor xa', 'formic acid', 'glutamyl endopeptidase', 'granzyme b', 'hydroxylamine', 'iodosobenzoic acid', 'lysc', 'lysn', 'ntcb', 'pepsin ph1.3', 'pepsin ph2.0', 'proline endopeptidase', 'proteinase k', 'staphylococcal peptidase i', 'thermolysin', 'thrombin', 'trypsin', 'trypsin_exception'] 
</p>

<p><strong>--cleavage-exception</strong> <strong><em>&lt;value&gt;</em></strong> <em>str</em><br/></p>
<p style="margin-left: 15px;">
Enzymatic cleavage exception.
 str <br/>Default: auto 

</p>

<p><strong>-m, --miscleavage</strong> <strong><em>&lt;number&gt;</em></strong> <em>int</em><br/></p>
<p style="margin-left: 15px;">
Number of cleavages to allow per non-canonical peptide.
 int <br/>Default: 2 

</p>

<p><strong>-w, --min-mw</strong> <strong><em>&lt;number&gt;</em></strong> <em>float</em><br/></p>
<p style="margin-left: 15px;">
The minimal molecular weight of the non-canonical peptides.
 float <br/>Default: 500.0 

</p>

<p><strong>-l, --min-length</strong> <strong><em>&lt;number&gt;</em></strong> <em>int</em><br/></p>
<p style="margin-left: 15px;">
The minimal length of non-canonical peptides, inclusive.
 int <br/>Default: 7 

</p>

<p><strong>-x, --max-length</strong> <strong><em>&lt;number&gt;</em></strong> <em>int</em><br/></p>
<p style="margin-left: 15px;">
The maximum length of non-canonical peptides, inclusive.
 int <br/>Default: 25 

</p>

<p><strong>--debug-level</strong> <strong><em>&lt;value|number&gt;</em></strong> <em>str</em><br/></p>
<p style="margin-left: 15px;">
Debug level.
 str <br/>Default: INFO 

</p>

<p><strong>-q, --quiet</strong><br/></p>
<p style="margin-left: 15px;">
Quiet
  <br/>Default: False 

</p>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../call-novel-orf/" class="btn btn-neutral float-right" title="callNovelORF">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../parse-circexplorer/" class="btn btn-neutral" title="parseCIRCexplorer"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/uclahs-cds/package-moPepGen/" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../parse-circexplorer/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../call-novel-orf/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme_extra.js" defer></script>
    <script src="../js/theme.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
