<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <meta name="author" content="Chenghao Zhu">
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>parseVEP - moPepGen</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="../css/theme.css" />
  <link rel="stylesheet" href="../css/theme_extra.css" />
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
  <link href="../assets/_mkdocstrings.css" rel="stylesheet" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "parseVEP";
    var mkdocs_page_input_path = "parse-vep.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> moPepGen</a>
        
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../basics/">moPepGen Basics</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../quick-start/">Quick Start</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../vignette/">Vignette</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../file-format/">File Format</a>
                    </li>
                </ul>
                <p class="caption"><span class="caption-text">Commands</span></p>
                <ul class="current">
                    <li class="toctree-l1"><a class="reference internal" href="../generate-index/">generateIndex</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../update-index/">updateIndex</a>
                    </li>
                    <li class="toctree-l1 current"><a class="reference internal current" href="./">parseVEP</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#moPepGen.cli.parse_vep">moPepGen.cli.parse_vep</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#usage">Usage</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#arguments">Arguments</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#running-vep-for-mopepgen">Running VEP for moPepGen</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#using-an-ensembl-gtf">Using an Ensembl GTF</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#using-a-non-ensembl-gtf">Using a Non-Ensembl GTF</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#tips-for-running-vep">Tips for Running VEP</a>
    </li>
        </ul>
    </li>
    </ul>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../parse-reditools/">parseREDItools</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../parse-star-fusion/">parseSTARFusion</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../parse-fusion-catcher/">parseFusionCatcher</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../parse-arriba/">parseArriba</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../parse-rmats/">parseRMATS</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../parse-circexplorer/">parseCIRCexplorer</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../call-variant/">callVariant</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../call-novel-orf/">callNovelORF</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../call-alt-translation/">callAltTranslation</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../filter-fasta/">filterFasta</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../split-fasta/">splitFasta</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../merge-fasta/">mergeFasta</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../encode-fasta/">encodeFasta</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../decoy-fasta/">decoyFasta</a>
                    </li>
                    <li class="toctree-l1"><a class="reference internal" href="../summarize-fasta/">summarizeFasta</a>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">moPepGen</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
        
          <li>Commands &raquo;</li>
        
      
    
    <li>parseVEP</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/uclahs-cds/package-moPepGen/edit/master/docs/parse-vep.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  
  <hr/>
</div>

          <div role="main">
            <div class="section">
              
                <h1 id="parsevep">parseVEP</h1>


  <div class="doc doc-object doc-module">

<a id="moPepGen.cli.parse_vep"></a>
    <div class="doc doc-contents first">

      <p><code>parseVEP</code> takes the output of Ensembl's <a href="https://uswest.ensembl.org/info/docs/tools/vep/index.html">Variant Effector Predictor
</a> (VEP) and
convert it into the GVF file format that moPepGen internally uses. The result
VEP file can then be parsed to moPepGen's <code>callVariant</code> subcommand to call for
variant peptide sequences.</p>


    </div>

  </div>

<div class="admonition note">
<p class="admonition-title">Reference Version</p>
<p>The version of reference genome and proteome FASTA and annotation GTF <strong>MUST</strong> be consistent across all analysis.</p>
</div>
<h2 id="usage">Usage</h2>
<pre><code>usage: moPepGen parseVEP [-h] -i ['&lt;files&gt;'] [['&lt;files&gt;'] ...] [-o &lt;file&gt;]
                         [--output-prefix &lt;value&gt;]
                         [--samples SAMPLES [SAMPLES ...]] --source SOURCE
                         [--skip-failed] [-g &lt;file&gt;] [-a &lt;file&gt;]
                         [--reference-source {GENCODE,ENSEMBL}]
                         [--codon-table {Standard,Alternative Flatworm Mitochondrial,Trematode Mitochondrial,SGC8,SGC3,Protozoan Mitochondrial,Ciliate Nuclear,Gracilibacteria,Spiroplasma,Dasycladacean Nuclear,Invertebrate Mitochondrial,Balanophoraceae Plastid,Peritrich Nuclear,Mesodinium Nuclear,SGC5,Candidate Division SR1,Blastocrithidia Nuclear,SGC1,Bacterial,Alternative Yeast Nuclear,Yeast Mitochondrial,Scenedesmus obliquus Mitochondrial,Plant Plastid,Flatworm Mitochondrial,SGC2,Archaeal,Mycoplasma,Euplotid Nuclear,SGC9,Mold Mitochondrial,Thraustochytrium Mitochondrial,Hexamita Nuclear,Coelenterate Mitochondrial,Chlorophycean Mitochondrial,Pachysolen tannophilus Nuclear,Ascidian Mitochondrial,SGC0,Blepharisma Macronuclear,Karyorelict Nuclear,SGC4,Echinoderm Mitochondrial,Condylostoma Nuclear,Vertebrate Mitochondrial,Pterobranchia Mitochondrial,Cephalodiscidae Mitochondrial}]
                         [--chr-codon-table [CHR_CODON_TABLE ...]]
                         [--start-codons [START_CODONS ...]]
                         [--chr-start-codons [CHR_START_CODONS ...]]
                         [--index-dir [&lt;file&gt;]] [--debug-level &lt;value|number&gt;]
                         [-q]

Parse VEP output tsv to the GVF format of variant records for moPepGen to call
variant peptides. The genome assembly FASTA and annotation GTF must come from
the same GENCODE/ENSEMBL version, and must the consistent with the VEP output.

options:
  -h, --help            show this help message and exit
  -i ['&lt;files&gt;'] [['&lt;files&gt;'] ...], --input-path ['&lt;files&gt;'] [['&lt;files&gt;'] ...]
                        File path to the VEP output TXT or VCF file. Can take
                        multiple files. Valid formats: ['.tsv', '.txt',
                        '.tsv.gz', '.txt.gz', '.vcf', '.vcf.gz'] (default:
                        None)
  -o &lt;file&gt;, --output-path &lt;file&gt;
                        File path to the output file. Valid formats: ['.gvf']
                        (default: None)
  --output-prefix &lt;value&gt;
                        Output prefix. Only used when inputs are VCF files.
                        (default: None)
  --samples SAMPLES [SAMPLES ...]
                        Samples to be parsed from the VCF file. If not
                        provided, all samples from the VCF file will be
                        parsed. This option is only used when the inputs are
                        VCF files. (default: [])
  --source SOURCE       Variant source (e.g. gSNP, sSNV, Fusion) (default:
                        None)
  --skip-failed         When set, the failed records will be skipped.
                        (default: False)
  --debug-level &lt;value|number&gt;
                        Debug level. (default: INFO)
  -q, --quiet           Quiet (default: False)

Reference Files:
  -g &lt;file&gt;, --genome-fasta &lt;file&gt;
                        Path to the genome assembly FASTA file. Only ENSEMBL
                        and GENCODE are supported. Its version must be the
                        same as the annotation GTF and proteome FASTA
                        (default: None)
  -a &lt;file&gt;, --annotation-gtf &lt;file&gt;
                        Path to the annotation GTF file. Only ENSEMBL and
                        GENCODE are supported. Its version must be the same as
                        the genome and proteome FASTA. (default: None)
  --reference-source {GENCODE,ENSEMBL}
                        Source of reference genome and annotation. (default:
                        None)
  --codon-table {Standard,Alternative Flatworm Mitochondrial,Trematode Mitochondrial,SGC8,SGC3,Protozoan Mitochondrial,Ciliate Nuclear,Gracilibacteria,Spiroplasma,Dasycladacean Nuclear,Invertebrate Mitochondrial,Balanophoraceae Plastid,Peritrich Nuclear,Mesodinium Nuclear,SGC5,Candidate Division SR1,Blastocrithidia Nuclear,SGC1,Bacterial,Alternative Yeast Nuclear,Yeast Mitochondrial,Scenedesmus obliquus Mitochondrial,Plant Plastid,Flatworm Mitochondrial,SGC2,Archaeal,Mycoplasma,Euplotid Nuclear,SGC9,Mold Mitochondrial,Thraustochytrium Mitochondrial,Hexamita Nuclear,Coelenterate Mitochondrial,Chlorophycean Mitochondrial,Pachysolen tannophilus Nuclear,Ascidian Mitochondrial,SGC0,Blepharisma Macronuclear,Karyorelict Nuclear,SGC4,Echinoderm Mitochondrial,Condylostoma Nuclear,Vertebrate Mitochondrial,Pterobranchia Mitochondrial,Cephalodiscidae Mitochondrial}
                        Codon table. Defaults to &quot;Standard&quot;. Supported codon
                        tables: {'Standard', 'Alternative Flatworm
                        Mitochondrial', 'Trematode Mitochondrial', 'SGC8',
                        'SGC3', 'Protozoan Mitochondrial', 'Ciliate Nuclear',
                        'Gracilibacteria', 'Spiroplasma', 'Dasycladacean
                        Nuclear', 'Invertebrate Mitochondrial',
                        'Balanophoraceae Plastid', 'Peritrich Nuclear',
                        'Mesodinium Nuclear', 'SGC5', 'Candidate Division
                        SR1', 'Blastocrithidia Nuclear', 'SGC1', 'Bacterial',
                        'Alternative Yeast Nuclear', 'Yeast Mitochondrial',
                        'Scenedesmus obliquus Mitochondrial', 'Plant Plastid',
                        'Flatworm Mitochondrial', 'SGC2', 'Archaeal',
                        'Mycoplasma', 'Euplotid Nuclear', 'SGC9', 'Mold
                        Mitochondrial', 'Thraustochytrium Mitochondrial',
                        'Hexamita Nuclear', 'Coelenterate Mitochondrial',
                        'Chlorophycean Mitochondrial', 'Pachysolen tannophilus
                        Nuclear', 'Ascidian Mitochondrial', 'SGC0',
                        'Blepharisma Macronuclear', 'Karyorelict Nuclear',
                        'SGC4', 'Echinoderm Mitochondrial', 'Condylostoma
                        Nuclear', 'Vertebrate Mitochondrial', 'Pterobranchia
                        Mitochondrial', 'Cephalodiscidae Mitochondrial'}
                        (default: Standard)
  --chr-codon-table [CHR_CODON_TABLE ...]
                        Chromosome specific codon table. Must be specified in
                        the format of &quot;chrM:SGC1&quot;, where &quot;chrM&quot; is the
                        chromosome name and &quot;SGC1&quot; is the codon table to use
                        to translate genes on chrM. Supported codon tables:
                        {'Standard', 'Alternative Flatworm Mitochondrial',
                        'Trematode Mitochondrial', 'SGC8', 'SGC3', 'Protozoan
                        Mitochondrial', 'Ciliate Nuclear', 'Gracilibacteria',
                        'Spiroplasma', 'Dasycladacean Nuclear', 'Invertebrate
                        Mitochondrial', 'Balanophoraceae Plastid', 'Peritrich
                        Nuclear', 'Mesodinium Nuclear', 'SGC5', 'Candidate
                        Division SR1', 'Blastocrithidia Nuclear', 'SGC1',
                        'Bacterial', 'Alternative Yeast Nuclear', 'Yeast
                        Mitochondrial', 'Scenedesmus obliquus Mitochondrial',
                        'Plant Plastid', 'Flatworm Mitochondrial', 'SGC2',
                        'Archaeal', 'Mycoplasma', 'Euplotid Nuclear', 'SGC9',
                        'Mold Mitochondrial', 'Thraustochytrium
                        Mitochondrial', 'Hexamita Nuclear', 'Coelenterate
                        Mitochondrial', 'Chlorophycean Mitochondrial',
                        'Pachysolen tannophilus Nuclear', 'Ascidian
                        Mitochondrial', 'SGC0', 'Blepharisma Macronuclear',
                        'Karyorelict Nuclear', 'SGC4', 'Echinoderm
                        Mitochondrial', 'Condylostoma Nuclear', 'Vertebrate
                        Mitochondrial', 'Pterobranchia Mitochondrial',
                        'Cephalodiscidae Mitochondrial'}. By default, &quot;SGC1&quot;
                        is assigned to mitochondrial chromosomes. (default:
                        [])
  --start-codons [START_CODONS ...]
                        Default start codon(s) to use for novel ORF
                        translation. Defaults to [&quot;ATG&quot;]. (default: ['ATG'])
  --chr-start-codons [CHR_START_CODONS ...]
                        Chromosome specific start codon(s). For example,
                        &quot;chrM:ATG,ATA,ATT&quot;.By defualt, mitochondrial
                        chromosome name is automatically inferred andstart
                        codon &quot;ATG&quot;, &quot;ATA&quot;, &quot;ATT&quot;, &quot;ATC&quot; and &quot;GTG&quot; are
                        assigned to it. (default: [])
  --index-dir [&lt;file&gt;]  Path to the directory of index files generated by
                        moPepGen generateIndex. If given, --genome-fasta,
                        --proteome-fasta and --anntotation-gtf will be
                        ignored. (default: None)

</code></pre>
<h2 id="arguments">Arguments</h2>
<p><strong>-h, --help</strong><br/></p>
<p style="margin-left: 15px;">
show this help message and exit


</p>

<p><strong>-i, --input-path</strong> <strong><em>[&#39;&lt;files&gt;&#39;]</em></strong> <em>Path</em><br/></p>
<p style="margin-left: 15px;">
File path to the VEP output TXT or VCF file. Can take multiple files. Valid formats: ['.tsv', '.txt', '.tsv.gz', '.txt.gz', '.vcf', '.vcf.gz']


</p>

<p><strong>-o, --output-path</strong> <strong><em>&lt;file&gt;</em></strong> <em>Path</em><br/></p>
<p style="margin-left: 15px;">
File path to the output file. Valid formats: ['.gvf']


</p>

<p><strong>--output-prefix</strong> <strong><em>&lt;value&gt;</em></strong> <em>Path</em><br/></p>
<p style="margin-left: 15px;">
Output prefix. Only used when inputs are VCF files.


</p>

<p><strong>--samples</strong> <em>str</em><br/></p>
<p style="margin-left: 15px;">
Samples to be parsed from the VCF file. If not provided, all samples from the VCF file will be parsed. This option is only used when the inputs are VCF files.
 str <br/>Default: [] 

</p>

<p><strong>--source</strong> <em>str</em><br/></p>
<p style="margin-left: 15px;">
Variant source (e.g. gSNP, sSNV, Fusion)


</p>

<p><strong>--skip-failed</strong><br/></p>
<p style="margin-left: 15px;">
When set, the failed records will be skipped.
  <br/>Default: False 

</p>

<p><strong>-g, --genome-fasta</strong> <strong><em>&lt;file&gt;</em></strong> <em>Path</em><br/></p>
<p style="margin-left: 15px;">
Path to the genome assembly FASTA file. Only ENSEMBL and GENCODE are supported. Its version must be the same as the annotation GTF and proteome FASTA


</p>

<p><strong>-a, --annotation-gtf</strong> <strong><em>&lt;file&gt;</em></strong> <em>Path</em><br/></p>
<p style="margin-left: 15px;">
Path to the annotation GTF file. Only ENSEMBL and GENCODE are supported. Its version must be the same as the genome and proteome FASTA.


</p>

<p><strong>--reference-source</strong> <em>str</em><br/></p>
<p style="margin-left: 15px;">
Source of reference genome and annotation.

 <br/>Choices: ['GENCODE', 'ENSEMBL'] 
</p>

<p><strong>--codon-table</strong> <em>str</em><br/></p>
<p style="margin-left: 15px;">
Codon table. Defaults to "Standard". Supported codon tables: {'Standard', 'Alternative Flatworm Mitochondrial', 'Trematode Mitochondrial', 'SGC8', 'SGC3', 'Protozoan Mitochondrial', 'Ciliate Nuclear', 'Gracilibacteria', 'Spiroplasma', 'Dasycladacean Nuclear', 'Invertebrate Mitochondrial', 'Balanophoraceae Plastid', 'Peritrich Nuclear', 'Mesodinium Nuclear', 'SGC5', 'Candidate Division SR1', 'Blastocrithidia Nuclear', 'SGC1', 'Bacterial', 'Alternative Yeast Nuclear', 'Yeast Mitochondrial', 'Scenedesmus obliquus Mitochondrial', 'Plant Plastid', 'Flatworm Mitochondrial', 'SGC2', 'Archaeal', 'Mycoplasma', 'Euplotid Nuclear', 'SGC9', 'Mold Mitochondrial', 'Thraustochytrium Mitochondrial', 'Hexamita Nuclear', 'Coelenterate Mitochondrial', 'Chlorophycean Mitochondrial', 'Pachysolen tannophilus Nuclear', 'Ascidian Mitochondrial', 'SGC0', 'Blepharisma Macronuclear', 'Karyorelict Nuclear', 'SGC4', 'Echinoderm Mitochondrial', 'Condylostoma Nuclear', 'Vertebrate Mitochondrial', 'Pterobranchia Mitochondrial', 'Cephalodiscidae Mitochondrial'}
 str <br/>Default: Standard 
 <br/>Choices: {'Standard', 'Alternative Flatworm Mitochondrial', 'Trematode Mitochondrial', 'SGC8', 'SGC3', 'Protozoan Mitochondrial', 'Ciliate Nuclear', 'Gracilibacteria', 'Spiroplasma', 'Dasycladacean Nuclear', 'Invertebrate Mitochondrial', 'Balanophoraceae Plastid', 'Peritrich Nuclear', 'Mesodinium Nuclear', 'SGC5', 'Candidate Division SR1', 'Blastocrithidia Nuclear', 'SGC1', 'Bacterial', 'Alternative Yeast Nuclear', 'Yeast Mitochondrial', 'Scenedesmus obliquus Mitochondrial', 'Plant Plastid', 'Flatworm Mitochondrial', 'SGC2', 'Archaeal', 'Mycoplasma', 'Euplotid Nuclear', 'SGC9', 'Mold Mitochondrial', 'Thraustochytrium Mitochondrial', 'Hexamita Nuclear', 'Coelenterate Mitochondrial', 'Chlorophycean Mitochondrial', 'Pachysolen tannophilus Nuclear', 'Ascidian Mitochondrial', 'SGC0', 'Blepharisma Macronuclear', 'Karyorelict Nuclear', 'SGC4', 'Echinoderm Mitochondrial', 'Condylostoma Nuclear', 'Vertebrate Mitochondrial', 'Pterobranchia Mitochondrial', 'Cephalodiscidae Mitochondrial'} 
</p>

<p><strong>--chr-codon-table</strong> <em>str</em><br/></p>
<p style="margin-left: 15px;">
Chromosome specific codon table. Must be specified in the format of "chrM:SGC1", where "chrM" is the chromosome name and "SGC1" is the codon table to use to translate genes on chrM. Supported codon tables: {'Standard', 'Alternative Flatworm Mitochondrial', 'Trematode Mitochondrial', 'SGC8', 'SGC3', 'Protozoan Mitochondrial', 'Ciliate Nuclear', 'Gracilibacteria', 'Spiroplasma', 'Dasycladacean Nuclear', 'Invertebrate Mitochondrial', 'Balanophoraceae Plastid', 'Peritrich Nuclear', 'Mesodinium Nuclear', 'SGC5', 'Candidate Division SR1', 'Blastocrithidia Nuclear', 'SGC1', 'Bacterial', 'Alternative Yeast Nuclear', 'Yeast Mitochondrial', 'Scenedesmus obliquus Mitochondrial', 'Plant Plastid', 'Flatworm Mitochondrial', 'SGC2', 'Archaeal', 'Mycoplasma', 'Euplotid Nuclear', 'SGC9', 'Mold Mitochondrial', 'Thraustochytrium Mitochondrial', 'Hexamita Nuclear', 'Coelenterate Mitochondrial', 'Chlorophycean Mitochondrial', 'Pachysolen tannophilus Nuclear', 'Ascidian Mitochondrial', 'SGC0', 'Blepharisma Macronuclear', 'Karyorelict Nuclear', 'SGC4', 'Echinoderm Mitochondrial', 'Condylostoma Nuclear', 'Vertebrate Mitochondrial', 'Pterobranchia Mitochondrial', 'Cephalodiscidae Mitochondrial'}. By default, "SGC1" is assigned to mitochondrial chromosomes.
 str <br/>Default: [] 

</p>

<p><strong>--start-codons</strong> <em>str</em><br/></p>
<p style="margin-left: 15px;">
Default start codon(s) to use for novel ORF translation. Defaults to ["ATG"].
 str <br/>Default: ['ATG'] 

</p>

<p><strong>--chr-start-codons</strong> <em>str</em><br/></p>
<p style="margin-left: 15px;">
Chromosome specific start codon(s). For example, "chrM:ATG,ATA,ATT".By defualt, mitochondrial chromosome name is automatically inferred andstart codon "ATG", "ATA", "ATT", "ATC" and "GTG" are assigned to it.
 str <br/>Default: [] 

</p>

<p><strong>--index-dir</strong> <strong><em>&lt;file&gt;</em></strong> <em>Path</em><br/></p>
<p style="margin-left: 15px;">
Path to the directory of index files generated by moPepGen generateIndex. If given, --genome-fasta, --proteome-fasta and --anntotation-gtf will be ignored.


</p>

<p><strong>--debug-level</strong> <strong><em>&lt;value|number&gt;</em></strong> <em>str</em><br/></p>
<p style="margin-left: 15px;">
Debug level.
 str <br/>Default: INFO 

</p>

<p><strong>-q, --quiet</strong><br/></p>
<p style="margin-left: 15px;">
Quiet
  <br/>Default: False 

</p>

<h2 id="running-vep-for-mopepgen">Running VEP for moPepGen</h2>
<p>moPepGen relies on Ensembl Variant Effect Predictor (<a href="https://www.ensembl.org/info/docs/tools/vep/index.html">VEP</a>) to annotate single nucleotide and small insertion and deletion (indel) variants, typically presented in VCF files. Note that VEP supports the annotation of multi-nucleotide variants and structural variants, but those are not currently supported by moPepGen and will be discarded by <code>parseVEP</code>.</p>
<p>Please refer to the official <a href="https://www.ensembl.org/info/docs/tools/vep/script/vep_tutorial.html">VEP Tutorial</a> for instructions on downloading, installing and running VEP. The key is to ensure that the annotation version used in <code>VEP</code> is the same as the one you would like to use with moPepGen.</p>
<p><code>parseVEP</code> supports both the TSV and VCF format of output of <code>VEP</code>. For TSV, <code>VEP</code>'s default is tab-delimited, and the <code>--tab</code> option adds extra columns to the output that are not used by moPepGen. So including or excluding <code>--tab</code> will not impact the results of moPepGen. As for <code>VCF</code>, the <code>CSQ</code> info field is parsed, and the default format is expected, containing the following fields:</p>
<pre><code>Allele|Consequence|IMPACT|SYMBOL|Gene|Feature_type|Feature|BIOTYPE|EXON|INTRON|HGVSc|HGVSp|cDNA_position|CDS_position|Protein_position|Amino_acids|Codons|Existing_variation|DISTANCE|STRAND|FLAGS|SYMBOL_SOURCE|HGNC_ID
</code></pre>
<p>When using VCF format, the <code>--samples</code> argument can be used to specify the columns you would like to parse, and the variant records will be split into sample-specific GVF files. By default, all sample columns will be processed.</p>
<h3 id="using-an-ensembl-gtf">Using an Ensembl GTF</h3>
<p><code>VEP</code> automatically uses an Ensembl GTF for annotation. Please ensure that the release version and genome build version is consistent with the references files downloaded. This is available in the name of the cache in output headers (for example, <code>107_GRCh38</code>).</p>
<p>One can ensure that the correct version is used by downloading the cache following <a href="https://www.ensembl.org/info/docs/tools/vep/script/vep_cache.html#cache">instructions</a> and specifying the parameters:
<code>--species homo_sapiens --assembly GRCh38 --cache --cache_version 107</code></p>
<h3 id="using-a-non-ensembl-gtf">Using a Non-Ensembl GTF</h3>
<p>If you have decided to use a set of reference files from GENCODE, it is important to supply VEP with the non-Ensembl GTF during annotation, so that chromosome names, transcript IDs and transcript coordinates match with your intended reference files. This can be specified in <code>VEP</code>.</p>
<p>As <a href="https://www.ensembl.org/info/docs/tools/vep/script/vep_cache.html#gff">instructed</a> by <code>VEP</code>, your GTF file must be sorted in chromosomal order and indexed.</p>
<pre><code>grep -v &quot;#&quot; PATH_TO_GTF | sort -k1,1 -k4,4n -k5,5n -t$'\t' | bgzip -c &gt; PATH_TO_GTF_GZ
tabix -p gff PATH_TO_GTF_GZ # gtf is not a tabix format option, gff works
</code></pre>
<p>To use the GTF for annotation, run VEP with the additional parameters</p>
<pre><code>--custom PATH_TO_GTF,GENCODE,gtf --fasta PATH_TO_GENOME_FA
</code></pre>
<p>Followed by VEP, we run the <code>filter_vep</code> command to subset outputs to only those annotated using the GENCODE GTF, since VEP automatically outputs both native (Ensembl) and custom annotations.</p>
<pre><code>filter_vep -i VEP_OUTPUT_TSV -o FILTERED_VEP_OUTPUT_TSV --filter Source = GENCODE
</code></pre>
<h3 id="tips-for-running-vep">Tips for Running VEP</h3>
<p>We recommend the following parameters for running VEP for run time optimization, please select appropriate settings for your system.</p>
<ul>
<li><code>--offline --cache</code></li>
<li><code>--no_stats</code></li>
<li><code>--fork</code></li>
<li><code>--buffer_size</code></li>
<li><code>--distance 0</code> - only annotation variants directly overlapping gene coordinates</li>
<li><code>--chr 1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,X,Y,MT</code> - focus on primary assembly</li>
<li><code>--no_intergenic</code> - ignore intergenic variants</li>
</ul>
<p>We do not recommend adding any additional annotation flags since the information is not utilized by moPepGen and increases run time.</p>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../parse-reditools/" class="btn btn-neutral float-right" title="parseREDItools">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../update-index/" class="btn btn-neutral" title="updateIndex"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/uclahs-cds/package-moPepGen/" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../update-index/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../parse-reditools/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme_extra.js" defer></script>
    <script src="../js/theme.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
